### ISSUE TITLE
May need to use lock in async Optimizer Wrapper

### ISSUE THREAD

**Issue by mhaoli** [sql-machine-learning/elasticdl#1255]
Currently, `OptimizerWrapper` is not thread-safe for embedding table. However, in async mode, many threads call `OptimizerWrapper.apply_gradients` without using lock. 

After implementing PS, we should discuss whether we need to add a lock here.

**Comment by chunyang-wen** [comment]
I think this kind of wrapper is a little confusing. Can we just provide our optimizers to users instead of silently using them? For example:

 ```python
from elasticdl.optimizer.adam import AdamOptimizer
```

This way we do not have to detect whether users have used the embedding layer. We just tell users they need to use our optimizers if they use certain functions like the embedding layer.

**Comment by mhaoli** [comment]
1. Whether or not we implement ElasticDL's optimizers, we need to discuss whether we need to use locks in optimizers.
2. Some users told us they want to use native TensorFlow api.
3. Through optimizer wrapper, we don't need to implement every optimizer in ElasticDL, which would cause a lot of repeat code.

**Comment by terrytangyuan** [comment]
Just a reference, Horovod uses:
> Wrap optimizer in `hvd.DistributedOptimizer`. The distributed optimizer delegates gradient computation to the original optimizer, averages gradients using allreduce or allgather, and then applies those averaged gradients.

which can be used as the following:
```
opt = tf.train.AdagradOptimizer(0.01 * hvd.size())
# Add Horovod Distributed Optimizer
opt = hvd.DistributedOptimizer(opt)
```

**Comment by mhaoli** [comment]
`OptimizerWrapper` in ElasticDL behaves like `hvd.DistributedOptimizer`. It also delegates gradient computation to the original optimizer and does following things in brief: 
* pull embedding vectors and slot variables from distributed kv store.
* call `apply_gradient` function of TensorFlow optimizer.
* push updated embedding vectors and slot variables to distributed kv store.

This issue aims to describe the following problem:
*In async mode, many threads wants to update embedding vectors at the same time. However, these embedding vectors are saved in distributed kv store. If we need to ensure that updates from all threads do not conflict, we need to add a lock here.*

Currently we use Redis as our distributed kv store, which complicates the code of `OptimizerWrapper`. After implementing PS, `OptimizerWrapper` will be refactored as a part of PS and the code can be greatly simplified. So if we decide to add a lock, we can implement this after finishing PS.

**Comment by workingloong** [comment]
"The entire optimizer is currently thread compatible, not thread-safe. The user needs to perform synchronization if necessary." https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer?version=stable#thread_compatibility_2
I think we need to use a lock for async Optimizer Wrapper to apply gradient.
